---
tags: [Import-b3c7]
title: Vireobulk_analysis
created: '2022-10-06T15:16:14.991Z'
modified: '2022-10-14T15:48:10.971Z'
---

- [Vireobulk_analysis](#vireobulk-analysis)
  * [installation & requirement](#installation---requirement)
  * [quick start](#quick-start)
  * [tag usage](#tag-usage)
- [annotation preprocessing](#annotation-preprocessing)
# Vireobulk_analysis
Vireobulk is the bulk RNA-seq deconvolution pipeline based on SNVs of mutiplexing pooled donors. this is the guideline 

## installation & requirement
Vireo-bulk's main function has been fully integrated into VireoSNP.
the core function of version 0.0.0 Vireo-bulk requires following package:

- pandas
- [VireoSNPv0.5.7](https://github.com/single-cell-genetics/vireo)

## quick start

```bash
python vireobulk.py -c CELLDATA -d DONOR_REFERENCE 
```
generate summary file for demutiplexed donor ratio and model.theta

## tag usage
```
--cellData -c 
```
the cell.vcf file of bulk RNAseq data that piled up by genotyping methods like samtools, and freebayes. However, for mixed bulk data , genotyping is unnecessary ,**[cellSNP-lite](https://github.com/single-cell-genetics/cellsnp-lite) is strongly recommeded** to generate this input data

```
--donorFile -d
```
the donor genotype reference file of donors exist in pooled bulk samples ,you could extract genotyping information 

**important**:you could use -R from samtools to avoid accidental issues in data input

```
--outDir -o 
```
Dirtectory for output files,defult is current workdir

### **optional**

```
--gene
```
when this flag is input , run gene level demutiplexing mode
  
  
```
  --notLearnGT", dest="force_learnGT", default=True,
```
If use, close learnGT mode for ASE in donor ratio demutiplex, default use learned mode


```
--annotation -a
```
annotation for SNPs and Genes" , this file can be generated by annoVAR, detailed infomration is listed [below](#annotation-preprocessing)




# annotation preprocessing


