{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## header file function (need to be script )\n",
    "import sys\n",
    "import numpy as np\n",
    "import vireoSNP\n",
    "from vireoSNP.utils.vireo_base import match\n",
    "from vireoSNP.utils.vcf_utils import load_VCF, write_VCF, parse_donor_GPb\n",
    "from vireoSNP.utils.vcf_utils import read_sparse_GeneINFO, GenoINFO_maker\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from vireoSNP.utils.io_utils import match_donor_VCF\n",
    "import re as re\n",
    "def VCF_to_sdf(bulk_data):\n",
    "    df_empty=pd.DataFrame(columns=[\"variants\",\"AD\",\"DP\",\"GT\"])\n",
    "    bulk_dat = read_sparse_GeneINFO(bulk_data['GenoINFO'], keys=['AD', 'DP'])\n",
    "    df_empty[\"variants\"]=bulk_data[\"variants\"]\n",
    "    df_empty[\"AD\"]=bulk_dat[\"AD\"].toarray()\n",
    "    df_empty[\"DP\"]=bulk_dat[\"DP\"].toarray()\n",
    "    return (df_empty)\n",
    "\n",
    "def rematch_vcf(bulk_data,donor_vcf):\n",
    "    df_bulk=VCF_to_sdf(bulk_data)\n",
    "    df_donor=pd.DataFrame(columns=[\"variants\",\"GT\"])\n",
    "    df_donor[\"variants\"]=donor_vcf[\"variants\"]\n",
    "    df_donor[\"GT\"]=donor_vcf[\"GenoINFO\"][\"GT\"]     \n",
    "    donor_tensor = vireoSNP.vcf.parse_donor_GPb(donor_vcf['GenoINFO'][\"GT\"], \"GT\")\n",
    "    itersect_variants=pd.Series(list(set(df_donor[\"variants\"]).intersection(set(df_bulk[\"variants\"]))))\n",
    "    id_d=df_donor[\"variants\"].isin(itersect_variants)\n",
    "    id_b=df_bulk[\"variants\"].isin(itersect_variants)\n",
    "    df_bulk=df_bulk[id_b]\n",
    "    df_donor=df_donor[id_d]\n",
    "    donor_tensor=donor_tensor[id_d]\n",
    "    return(df_bulk,df_donor,donor_tensor)\n",
    "def rematch_vcf_annotated(bulk_data,donor_vcf,donor_vcf_anno):\n",
    "    df_bulk=VCF_to_sdf(bulk_data)\n",
    "    df_donora=pd.DataFrame(columns=[\"variants\",\"function\",\"genes\",\"avsnp\",\"GT\"])\n",
    "    df_donora[\"variants\"]=donor_vcf_anno[\"variants\"]  \n",
    "    df_donora[\"function\"]=donor_vcf_anno[\"function.\"]\n",
    "    df_donora[\"genes\"]=donor_vcf_anno[\"genes\"]\n",
    "    df_donora[\"avsnp\"]=donor_vcf_anno[\"avsnp\"]\n",
    "    df_donor=pd.DataFrame(columns=[\"variants\",\"GT\"])\n",
    "    df_donor[\"variants\"]=donor_vcf[\"variants\"]\n",
    "    df_donor[\"GT\"]=donor_vcf[\"GenoINFO\"][\"GT\"]     \n",
    "    donor_tensor = vireoSNP.vcf.parse_donor_GPb(donor_vcf['GenoINFO'][\"GT\"], \"GT\")\n",
    "    itersect_variants=pd.Series(list(set(df_donora[\"variants\"]).intersection(set(df_bulk[\"variants\"]))))\n",
    "    id_d=df_donor[\"variants\"].isin(itersect_variants)\n",
    "    id_b=df_bulk[\"variants\"].isin(itersect_variants)\n",
    "    id_da=df_donora[\"variants\"].isin(itersect_variants)\n",
    "    df_bulk=df_bulk[id_b]\n",
    "    df_donor=df_donor[id_d]\n",
    "    df_donora=df_donora[id_da]\n",
    "    donor_tensor=donor_tensor[id_d]\n",
    "    df_donora[\"GT\"]=df_donora[\"GT\"]=df_donor[\"GT\"].tolist()\n",
    "    return(df_bulk,df_donora,donor_tensor)\n",
    "def merged_VCF_to_sdf(bulk_data_merge):\n",
    "    df_empty=pd.DataFrame(columns=[\"variants\",\"AD\",\"DP\",\"OTH\"])\n",
    "    df_empty[\"variants\"]=bulk_data_merge[\"variants\"]\n",
    "    testdf=pd.DataFrame(np.array(bulk_merged_data['FixedINFO'][\"INFO\"]))\n",
    "    df_empty[\"AD\"]=testdf[0].str.split(';',expand=True)[0].str.split('=',expand=True)[1].astype(float)\n",
    "    df_empty[\"DP\"]=testdf[0].str.split(';',expand=True)[1].str.split('=',expand=True)[1].astype(float)\n",
    "    df_empty[\"OTH\"]=testdf[0].str.split(';',expand=True)[2].str.split('=',expand=True)[1].astype(float)\n",
    "\n",
    "    \n",
    "\n",
    "    return(df_empty)\n",
    "def rematch_merge_vcf_annotated(bulk_data,donor_vcf,donor_vcf_anno):\n",
    "    df_bulk=merged_VCF_to_sdf(bulk_data)\n",
    "    df_donora=pd.DataFrame(columns=[\"variants\",\"function\",\"genes\",\"avsnp\",\"GT\"])\n",
    "    df_donora[\"variants\"]=donor_vcf_anno[\"variants\"]  \n",
    "    df_donora[\"function\"]=donor_vcf_anno[\"function.\"]\n",
    "    df_donora[\"genes\"]=donor_vcf_anno[\"genes\"]\n",
    "    df_donora[\"avsnp\"]=donor_vcf_anno[\"avsnp\"]\n",
    "    df_donor=pd.DataFrame(columns=[\"variants\",\"GT\"])\n",
    "    df_donor[\"variants\"]=donor_vcf[\"variants\"]\n",
    "    df_donor[\"GT\"]=donor_vcf[\"GenoINFO\"][\"GT\"]     \n",
    "    donor_tensor = vireoSNP.vcf.parse_donor_GPb(donor_vcf['GenoINFO'][\"GT\"], \"GT\")\n",
    "    itersect_variants=pd.Series(list(set(df_donora[\"variants\"]).intersection(set(df_bulk[\"variants\"]))))\n",
    "    id_d=df_donor[\"variants\"].isin(itersect_variants)\n",
    "    id_b=df_bulk[\"variants\"].isin(itersect_variants)\n",
    "    id_da=df_donora[\"variants\"].isin(itersect_variants)\n",
    "    df_bulk=df_bulk[id_b]\n",
    "    df_donor=df_donor[id_d]\n",
    "    df_donora=df_donora[id_da]\n",
    "    donor_tensor=donor_tensor[id_d]\n",
    "    df_donora[\"GT\"]=df_donora[\"GT\"]=df_donor[\"GT\"].tolist()\n",
    "    return(df_bulk,df_donora,donor_tensor)\n",
    "def find_all(data, s):\n",
    "    r_list = []\n",
    "    for r in range(len(data)):\n",
    "        r_list.append( s in data[r])\n",
    " \n",
    "    return(r_list)\n",
    "def removedfault(df_bulk,df_donora,donor_tensor):\n",
    "    id_d=[]\n",
    "    for bool in find_all(np.array(df_donora[\"GT\"]).tolist(),\"./.\"):\n",
    "        id_d.append(not bool)\n",
    "    df_donor=df_donora[id_d]\n",
    "    df_gt=donor_tensor[id_d]\n",
    "    id_b=df_bulk[\"variants\"].isin(df_donor[\"variants\"])\n",
    "    df_b=df_bulk[id_b]\n",
    "    return(df_b,df_donor,df_gt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prediction_bulk(df_b,df_d,df_gt,donor_vcf,pre_model,genelist):\n",
    "    gene_unique=genelist\n",
    "    new=pd.DataFrame(index=gene_unique,columns=donor_vcf['samples']+[\"chi\",\"p\",\"numSNP\"])\n",
    "    pre_psi=pre_model\n",
    "    test_d=df_d\n",
    "    test_b=df_b\n",
    "    test_gt=df_gt\n",
    "    sample_list=donor_vcf['samples']\n",
    "    nk=len(donor_vcf['samples'])\n",
    "    for i in range(len(gene_unique)):\n",
    "        sub_d=test_d[test_d[\"genes\"]==gene_unique[i]]\n",
    "        sub_b=test_b[test_b[\"variants\"].isin(test_d[test_d[\"genes\"]==gene_unique[i]][\"variants\"])]\n",
    "        sub_gt=test_gt[test_d[\"genes\"]==gene_unique[i]]\n",
    "        a=np.array(sub_b[\"AD\"])\n",
    "        d=np.array(sub_b[\"DP\"])\n",
    "        a=np.array(list(map(float,a)))\n",
    "        d=np.array(list(map(float,d)))\n",
    "        model =vireoSNP.VireoBulk(nk)\n",
    "        model.fit(a,d,sub_gt,learn_theta=False )\n",
    "        pv=vireoSNP.LikRatio_test(model.psi,pre_psi,a,d,sub_gt,model.theta)\n",
    "        for j in range(len(model.psi)):\n",
    "            new[sample_list[j]][i]=model.psi[j]\n",
    "        new[\"chi\"][i]=pv[0]\n",
    "        new[\"p\"][i]=pv[1]\n",
    "        new[\"numSNP\"][i]=len(a)\n",
    "    return(new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_vcf=load_VCF(   \"/home/flyflyzhizhi/Vireobulk_analysis/data/filter_pbmc10donors.vcf.gz\",    biallelic_only=True,    load_sample=True, sparse=False , format_list=None)\n",
    "####count based model\n",
    "##bulk_merged_data=load_VCF(   \"/media/wgs/subset_stimulation_bam/pbmc10donor_bulk/cellSNP.cells.vcf.gz\",    biallelic_only=True,    load_sample=True,    format_list=None,)\n",
    "#### UMI based model\n",
    "bulk_merged_data=load_VCF(\"/home/flyflyzhizhi/Vireobulk_analysis/data/cellSNP.base.vcf.gz\", biallelic_only=True,load_sample=False,format_list=None,)\n",
    "###processedcsv\n",
    "df_anno=pd.read_csv(\"/home/flyflyzhizhi/Vireobulk_analysis/data/scvcf_annotated_processed.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulk_merged_data.t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variants</th>\n",
       "      <th>AD</th>\n",
       "      <th>DP</th>\n",
       "      <th>OTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_14574_A_G</td>\n",
       "      <td>2.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_14590_G_A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_14599_T_A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_14604_A_G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_14610_T_C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128194</th>\n",
       "      <td>chrY_13390522_T_C</td>\n",
       "      <td>66.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128195</th>\n",
       "      <td>chrY_19413646_C_T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128196</th>\n",
       "      <td>chrY_19413653_G_A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128197</th>\n",
       "      <td>chrY_19604311_A_T</td>\n",
       "      <td>45.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128198</th>\n",
       "      <td>chrY_20577481_C_T</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128199 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 variants    AD    DP  OTH\n",
       "0          chr1_14574_A_G   2.0  35.0  1.0\n",
       "1          chr1_14590_G_A   0.0  50.0  0.0\n",
       "2          chr1_14599_T_A   0.0  53.0  2.0\n",
       "3          chr1_14604_A_G   1.0  60.0  0.0\n",
       "4          chr1_14610_T_C   1.0  62.0  1.0\n",
       "...                   ...   ...   ...  ...\n",
       "128194  chrY_13390522_T_C  66.0  66.0  0.0\n",
       "128195  chrY_19413646_C_T   0.0  20.0  0.0\n",
       "128196  chrY_19413653_G_A   0.0  20.0  0.0\n",
       "128197  chrY_19604311_A_T  45.0  45.0  0.0\n",
       "128198  chrY_20577481_C_T  30.0  30.0  0.0\n",
       "\n",
       "[128199 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.34 s, sys: 7.52 ms, total: 3.34 s\n",
      "Wall time: 3.34 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "df_b=merged_VCF_to_sdf(bulk_merged_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'GenoINFO'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_8831/3933178933.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_b\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdf_d\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mm_GT\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrematch_vcf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbulk_merged_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdonor_vcf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/tmp/ipykernel_8831/4286517543.py\u001b[0m in \u001b[0;36mrematch_vcf\u001b[0;34m(bulk_data, donor_vcf)\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mrematch_vcf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbulk_data\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdonor_vcf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m     \u001b[0mdf_bulk\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mVCF_to_sdf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbulk_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m     \u001b[0mdf_donor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"variants\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"GT\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0mdf_donor\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"variants\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdonor_vcf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"variants\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_8831/4286517543.py\u001b[0m in \u001b[0;36mVCF_to_sdf\u001b[0;34m(bulk_data)\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mVCF_to_sdf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbulk_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[0mdf_empty\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"variants\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"AD\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"DP\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"GT\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m     \u001b[0mbulk_dat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mread_sparse_GeneINFO\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbulk_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'GenoINFO'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'AD'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'DP'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m     \u001b[0mdf_empty\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"variants\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbulk_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"variants\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m     \u001b[0mdf_empty\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"AD\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbulk_dat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"AD\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'GenoINFO'"
     ]
    }
   ],
   "source": [
    "df_b,df_d,m_GT=rematch_vcf(bulk_merged_data, donor_vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count_bulk_data\n",
    "count_bulk_data=load_VCF(   \"/media/wgs/subset_stimulation_bam/pbmc10donor_bulk/cellSNP.cells.vcf.gz\",    biallelic_only=True,    load_sample=True,    format_list=None,)\n",
    "cf_b,cf_d,cm_GT=rematch_merge_vcf_annotated(count_bulk_data,donor_vcf,df_anno)\n",
    "cf_b,cf_d,cm_GT=removedfault(cf_b,cf_d,cm_GT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_tensor = vireoSNP.vcf.parse_donor_GPb(donor_vcf['GenoINFO'][\"GT\"], \"GT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[1.        , 0.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        ...,\n",
       "        [1.        , 0.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ]],\n",
       "\n",
       "       [[1.        , 0.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        ...,\n",
       "        [1.        , 0.        , 0.        ],\n",
       "        [1.        , 0.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ]],\n",
       "\n",
       "       [[0.        , 1.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        ...,\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ]],\n",
       "\n",
       "       ...,\n",
       "\n",
       "       [[0.        , 1.        , 0.        ],\n",
       "        [1.        , 0.        , 0.        ],\n",
       "        [1.        , 0.        , 0.        ],\n",
       "        ...,\n",
       "        [1.        , 0.        , 0.        ],\n",
       "        [0.        , 1.        , 0.        ],\n",
       "        [1.        , 0.        , 0.        ]],\n",
       "\n",
       "       [[0.        , 0.        , 1.        ],\n",
       "        [0.33333333, 0.33333333, 0.33333333],\n",
       "        [0.        , 0.        , 1.        ],\n",
       "        ...,\n",
       "        [0.        , 0.        , 1.        ],\n",
       "        [0.        , 0.        , 1.        ],\n",
       "        [0.33333333, 0.33333333, 0.33333333]],\n",
       "\n",
       "       [[0.        , 0.        , 1.        ],\n",
       "        [0.33333333, 0.33333333, 0.33333333],\n",
       "        [0.        , 0.        , 1.        ],\n",
       "        ...,\n",
       "        [0.        , 0.        , 1.        ],\n",
       "        [0.        , 0.        , 1.        ],\n",
       "        [0.33333333, 0.33333333, 0.33333333]]])"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "donor_tensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "VCF_to_sdf(bulk_merged_data,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.64 s, sys: 12 ms, total: 5.65 s\n",
      "Wall time: 5.65 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "df_b,df_d,m_GT=rematch_merge_vcf_annotated(bulk_merged_data,donor_vcf,df_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variants</th>\n",
       "      <th>function</th>\n",
       "      <th>genes</th>\n",
       "      <th>avsnp</th>\n",
       "      <th>GT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_14574_A_G</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs28503599</td>\n",
       "      <td>[0/0, 0/1, 0/1, 0/0, 0/0, 0/0, 0/1, 0/0, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_14590_G_A</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs707679</td>\n",
       "      <td>[0/0, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/0, 0/0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_14599_T_A</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs707680</td>\n",
       "      <td>[0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_14604_A_G</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs541940975</td>\n",
       "      <td>[0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_14610_T_C</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs878986575</td>\n",
       "      <td>[0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128180</th>\n",
       "      <td>chrX_155999818_C_T</td>\n",
       "      <td>intronic</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>rs3093470</td>\n",
       "      <td>[1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128181</th>\n",
       "      <td>chrX_155999869_A_G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>rs3091261</td>\n",
       "      <td>[1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128182</th>\n",
       "      <td>chrX_155999957_G_C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>rs3093472</td>\n",
       "      <td>[1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128183</th>\n",
       "      <td>chrX_156025297_G_A</td>\n",
       "      <td>downstream</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>rs185932868</td>\n",
       "      <td>[0/1, 0/0, 0/1, 0/0, 0/1, 0/0, 0/0, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128187</th>\n",
       "      <td>chrY_3120449_C_G</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LINC00278;TGIF2LY</td>\n",
       "      <td>rs778218266</td>\n",
       "      <td>[0/1, 1/1, 0/1, 1/1, 0/0, 1/1, 1/1, 0/0, 0/1, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127155 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variants      function              genes        avsnp  \\\n",
       "0           chr1_14574_A_G  ncRNA_exonic             WASH7P   rs28503599   \n",
       "1           chr1_14590_G_A  ncRNA_exonic             WASH7P     rs707679   \n",
       "2           chr1_14599_T_A  ncRNA_exonic             WASH7P     rs707680   \n",
       "3           chr1_14604_A_G  ncRNA_exonic             WASH7P  rs541940975   \n",
       "4           chr1_14610_T_C  ncRNA_exonic             WASH7P  rs878986575   \n",
       "...                    ...           ...                ...          ...   \n",
       "128180  chrX_155999818_C_T      intronic               IL9R    rs3093470   \n",
       "128181  chrX_155999869_A_G      intronic               IL9R    rs3091261   \n",
       "128182  chrX_155999957_G_C      intronic               IL9R    rs3093472   \n",
       "128183  chrX_156025297_G_A    downstream           DDX11L16  rs185932868   \n",
       "128187    chrY_3120449_C_G    intergenic  LINC00278;TGIF2LY  rs778218266   \n",
       "\n",
       "                                                       GT  \n",
       "0       [0/0, 0/1, 0/1, 0/0, 0/0, 0/0, 0/1, 0/0, 0/1, ...  \n",
       "1       [0/0, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/0, 0/0, ...  \n",
       "2       [0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...  \n",
       "3       [0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...  \n",
       "4       [0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...  \n",
       "...                                                   ...  \n",
       "128180  [1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...  \n",
       "128181  [1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, ...  \n",
       "128182  [1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...  \n",
       "128183  [0/1, 0/0, 0/1, 0/0, 0/1, 0/0, 0/0, 0/1, 0/1, ...  \n",
       "128187  [0/1, 1/1, 0/1, 1/1, 0/0, 1/1, 1/1, 0/0, 0/1, ...  \n",
       "\n",
       "[127155 rows x 5 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_b,df_d,m_GT=removedfault(df_b,df_d,m_GT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(127155, 10, 3)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_GT.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_unique=np.unique(df_d[\"genes\"].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=np.array(df_b[\"AD\"])\n",
    "d=np.array(df_b[\"DP\"])\n",
    "a=np.array(list(map(float,a)))\n",
    "d=np.array(list(map(float,d)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "model =vireoSNP.VireoBulk(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "model.fit(a,d,m_GT,learn_theta=True )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.01054732, 0.44960107, 0.99271081])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.theta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['18-G-017_R1',\n",
       " '18-G-018_R1',\n",
       " '18-G-019_R1',\n",
       " '18-G-020_R1',\n",
       " '18-G-021_R1',\n",
       " '18-G-022_R1',\n",
       " '18-G-023_R1',\n",
       " '18-G-024_R1',\n",
       " '18-G-025_R1',\n",
       " '18-G-026_R1']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "donor_vcf['samples']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "premodel=model.psi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.05715795, 0.04111812, 0.06364389, 0.09476673, 0.14070468,\n",
       "       0.07099676, 0.05302997, 0.11330121, 0.28828329, 0.07699739])"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "premodel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6min 7s, sys: 23.9 ms, total: 6min 7s\n",
      "Wall time: 6min 5s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "bulk_demuti_result= prediction_bulk(df_b,df_d,m_GT,donor_vcf,premodel,gene_unique)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>18-G-017_R1</th>\n",
       "      <th>18-G-018_R1</th>\n",
       "      <th>18-G-019_R1</th>\n",
       "      <th>18-G-020_R1</th>\n",
       "      <th>18-G-021_R1</th>\n",
       "      <th>18-G-022_R1</th>\n",
       "      <th>18-G-023_R1</th>\n",
       "      <th>18-G-024_R1</th>\n",
       "      <th>18-G-025_R1</th>\n",
       "      <th>18-G-026_R1</th>\n",
       "      <th>chi</th>\n",
       "      <th>p</th>\n",
       "      <th>numSNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.026748</td>\n",
       "      <td>0.062549</td>\n",
       "      <td>0.102357</td>\n",
       "      <td>0.022211</td>\n",
       "      <td>0.450597</td>\n",
       "      <td>0.027805</td>\n",
       "      <td>0.025727</td>\n",
       "      <td>0.101977</td>\n",
       "      <td>0.163915</td>\n",
       "      <td>0.016113</td>\n",
       "      <td>3.248674</td>\n",
       "      <td>0.953614</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.061176</td>\n",
       "      <td>0.010409</td>\n",
       "      <td>0.001743</td>\n",
       "      <td>0.049935</td>\n",
       "      <td>0.568557</td>\n",
       "      <td>0.097229</td>\n",
       "      <td>0.006311</td>\n",
       "      <td>0.011423</td>\n",
       "      <td>0.005862</td>\n",
       "      <td>0.187353</td>\n",
       "      <td>2.903783</td>\n",
       "      <td>0.967988</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.124275</td>\n",
       "      <td>0.001151</td>\n",
       "      <td>0.51316</td>\n",
       "      <td>0.09343</td>\n",
       "      <td>0.001208</td>\n",
       "      <td>0.008507</td>\n",
       "      <td>0.005611</td>\n",
       "      <td>0.113512</td>\n",
       "      <td>0.116549</td>\n",
       "      <td>0.022598</td>\n",
       "      <td>6.817806</td>\n",
       "      <td>0.656082</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M;PZP</th>\n",
       "      <td>0.170696</td>\n",
       "      <td>0.003699</td>\n",
       "      <td>0.326753</td>\n",
       "      <td>0.05438</td>\n",
       "      <td>0.042315</td>\n",
       "      <td>0.016654</td>\n",
       "      <td>0.085925</td>\n",
       "      <td>0.183231</td>\n",
       "      <td>0.113208</td>\n",
       "      <td>0.003139</td>\n",
       "      <td>5.912048</td>\n",
       "      <td>0.748688</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2MP1</th>\n",
       "      <td>0.059267</td>\n",
       "      <td>0.157766</td>\n",
       "      <td>0.088763</td>\n",
       "      <td>0.342016</td>\n",
       "      <td>0.023697</td>\n",
       "      <td>0.077471</td>\n",
       "      <td>0.190773</td>\n",
       "      <td>0.041997</td>\n",
       "      <td>0.018249</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>22.83156</td>\n",
       "      <td>0.006586</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>0.212285</td>\n",
       "      <td>0.038627</td>\n",
       "      <td>0.028323</td>\n",
       "      <td>0.213397</td>\n",
       "      <td>0.070399</td>\n",
       "      <td>0.005996</td>\n",
       "      <td>0.136235</td>\n",
       "      <td>0.143128</td>\n",
       "      <td>0.089694</td>\n",
       "      <td>0.061916</td>\n",
       "      <td>2.144412</td>\n",
       "      <td>0.988972</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.03371</td>\n",
       "      <td>0.019217</td>\n",
       "      <td>0.103537</td>\n",
       "      <td>0.342751</td>\n",
       "      <td>0.120287</td>\n",
       "      <td>0.01292</td>\n",
       "      <td>0.01835</td>\n",
       "      <td>0.13429</td>\n",
       "      <td>0.046355</td>\n",
       "      <td>0.168583</td>\n",
       "      <td>10.483539</td>\n",
       "      <td>0.312775</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0.026706</td>\n",
       "      <td>0.009975</td>\n",
       "      <td>0.145311</td>\n",
       "      <td>0.043494</td>\n",
       "      <td>0.113605</td>\n",
       "      <td>0.078055</td>\n",
       "      <td>0.0417</td>\n",
       "      <td>0.003031</td>\n",
       "      <td>0.406706</td>\n",
       "      <td>0.131415</td>\n",
       "      <td>5.698745</td>\n",
       "      <td>0.769649</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.012618</td>\n",
       "      <td>0.059708</td>\n",
       "      <td>0.149221</td>\n",
       "      <td>0.054568</td>\n",
       "      <td>0.227686</td>\n",
       "      <td>0.048068</td>\n",
       "      <td>0.067843</td>\n",
       "      <td>0.118152</td>\n",
       "      <td>0.161034</td>\n",
       "      <td>0.101103</td>\n",
       "      <td>1.681484</td>\n",
       "      <td>0.995563</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.0496</td>\n",
       "      <td>0.055932</td>\n",
       "      <td>0.111258</td>\n",
       "      <td>0.156045</td>\n",
       "      <td>0.055213</td>\n",
       "      <td>0.085981</td>\n",
       "      <td>0.034441</td>\n",
       "      <td>0.085136</td>\n",
       "      <td>0.192274</td>\n",
       "      <td>0.174121</td>\n",
       "      <td>3.636167</td>\n",
       "      <td>0.933688</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13350 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         18-G-017_R1 18-G-018_R1 18-G-019_R1 18-G-020_R1 18-G-021_R1  \\\n",
       "A1BG        0.026748    0.062549    0.102357    0.022211    0.450597   \n",
       "A1BG-AS1    0.061176    0.010409    0.001743    0.049935    0.568557   \n",
       "A2M         0.124275    0.001151     0.51316     0.09343    0.001208   \n",
       "A2M;PZP     0.170696    0.003699    0.326753     0.05438    0.042315   \n",
       "A2MP1       0.059267    0.157766    0.088763    0.342016    0.023697   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZXDC        0.212285    0.038627    0.028323    0.213397    0.070399   \n",
       "ZYG11B       0.03371    0.019217    0.103537    0.342751    0.120287   \n",
       "ZYX         0.026706    0.009975    0.145311    0.043494    0.113605   \n",
       "ZZEF1       0.012618    0.059708    0.149221    0.054568    0.227686   \n",
       "ZZZ3          0.0496    0.055932    0.111258    0.156045    0.055213   \n",
       "\n",
       "         18-G-022_R1 18-G-023_R1 18-G-024_R1 18-G-025_R1 18-G-026_R1  \\\n",
       "A1BG        0.027805    0.025727    0.101977    0.163915    0.016113   \n",
       "A1BG-AS1    0.097229    0.006311    0.011423    0.005862    0.187353   \n",
       "A2M         0.008507    0.005611    0.113512    0.116549    0.022598   \n",
       "A2M;PZP     0.016654    0.085925    0.183231    0.113208    0.003139   \n",
       "A2MP1       0.077471    0.190773    0.041997    0.018249    0.000001   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZXDC        0.005996    0.136235    0.143128    0.089694    0.061916   \n",
       "ZYG11B       0.01292     0.01835     0.13429    0.046355    0.168583   \n",
       "ZYX         0.078055      0.0417    0.003031    0.406706    0.131415   \n",
       "ZZEF1       0.048068    0.067843    0.118152    0.161034    0.101103   \n",
       "ZZZ3        0.085981    0.034441    0.085136    0.192274    0.174121   \n",
       "\n",
       "                chi         p numSNP  \n",
       "A1BG       3.248674  0.953614      3  \n",
       "A1BG-AS1   2.903783  0.967988      2  \n",
       "A2M        6.817806  0.656082      2  \n",
       "A2M;PZP    5.912048  0.748688      1  \n",
       "A2MP1      22.83156  0.006586      3  \n",
       "...             ...       ...    ...  \n",
       "ZXDC       2.144412  0.988972     13  \n",
       "ZYG11B    10.483539  0.312775      9  \n",
       "ZYX        5.698745  0.769649      2  \n",
       "ZZEF1      1.681484  0.995563     29  \n",
       "ZZZ3       3.636167  0.933688     25  \n",
       "\n",
       "[13350 rows x 13 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulk_demuti_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulk_demuti_result.to_csv(\"/media/wgs/subset_stimulation_bam/pbmc10donor/annotated_expo_umi2.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
