{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## header file function (need to be script )\n",
    "import sys\n",
    "import numpy as np\n",
    "import vireoSNP\n",
    "from vireoSNP.utils.vireo_base import match\n",
    "from vireoSNP.utils.vcf_utils import load_VCF, write_VCF, parse_donor_GPb\n",
    "from vireoSNP.utils.vcf_utils import read_sparse_GeneINFO, GenoINFO_maker\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from vireoSNP.utils.io_utils import match_donor_VCF\n",
    "import re as re\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/home/ccheng/Vireobulk_analysis/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from base import merged_VCF_to_sdf,rematch_merge_vcf_annotated,find_all,removedfault,prediction_bulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "##loadfile\n",
    "\n",
    "donor_vcf=load_VCF(   \"/media/wgs/subset_stimulation_bam/filter_pbmc10donors.vcf.gz\",    biallelic_only=True,    load_sample=True, sparse=False , format_list=None)\n",
    "####count based model\n",
    "##bulk_merged_data=load_VCF(   \"/media/wgs/subset_stimulation_bam/pbmc10donor_bulk/cellSNP.cells.vcf.gz\",    biallelic_only=True,    load_sample=True,    format_list=None,)\n",
    "#### UMI based model\n",
    "bulk_merged_data=load_VCF(\"/media/wgs/subset_stimulation_bam/10pbmcdonor/cellSNP.base.vcf.gz\", biallelic_only=True,load_sample=False,format_list=None,)\n",
    "###processedcsv\n",
    "df_anno=pd.read_csv(\"/media/wgs/subset_stimulation_bam/10pbmcdonor/pbmcvcf_annotated_processed.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#data clean\n",
    "df_b,df_d,m_GT=rematch_merge_vcf_annotated(bulk_merged_data,donor_vcf,df_anno)\n",
    "df_b,df_d,m_GT=removedfault(df_b,df_d,m_GT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variants</th>\n",
       "      <th>AD</th>\n",
       "      <th>DP</th>\n",
       "      <th>OTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_14574_A_G</td>\n",
       "      <td>2.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_14590_G_A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_14599_T_A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_14604_A_G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_14610_T_C</td>\n",
       "      <td>1.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128180</th>\n",
       "      <td>chrX_155999818_C_T</td>\n",
       "      <td>66.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128181</th>\n",
       "      <td>chrX_155999869_A_G</td>\n",
       "      <td>88.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128182</th>\n",
       "      <td>chrX_155999957_G_C</td>\n",
       "      <td>73.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128183</th>\n",
       "      <td>chrX_156025297_G_A</td>\n",
       "      <td>65.0</td>\n",
       "      <td>195.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128187</th>\n",
       "      <td>chrY_3120449_C_G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127155 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variants    AD     DP  OTH\n",
       "0           chr1_14574_A_G   2.0   35.0  1.0\n",
       "1           chr1_14590_G_A   0.0   50.0  0.0\n",
       "2           chr1_14599_T_A   0.0   53.0  2.0\n",
       "3           chr1_14604_A_G   1.0   60.0  0.0\n",
       "4           chr1_14610_T_C   1.0   62.0  1.0\n",
       "...                    ...   ...    ...  ...\n",
       "128180  chrX_155999818_C_T  66.0   80.0  0.0\n",
       "128181  chrX_155999869_A_G  88.0   89.0  0.0\n",
       "128182  chrX_155999957_G_C  73.0   86.0  0.0\n",
       "128183  chrX_156025297_G_A  65.0  195.0  1.0\n",
       "128187    chrY_3120449_C_G   0.0   28.0  1.0\n",
       "\n",
       "[127155 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_b\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variants</th>\n",
       "      <th>function</th>\n",
       "      <th>genes</th>\n",
       "      <th>avsnp</th>\n",
       "      <th>GT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_14574_A_G</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs28503599</td>\n",
       "      <td>[0/0, 0/1, 0/1, 0/0, 0/0, 0/0, 0/1, 0/0, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_14590_G_A</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs707679</td>\n",
       "      <td>[0/0, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/0, 0/0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_14599_T_A</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs707680</td>\n",
       "      <td>[0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_14604_A_G</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs541940975</td>\n",
       "      <td>[0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_14610_T_C</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>rs878986575</td>\n",
       "      <td>[0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128180</th>\n",
       "      <td>chrX_155999818_C_T</td>\n",
       "      <td>intronic</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>rs3093470</td>\n",
       "      <td>[1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128181</th>\n",
       "      <td>chrX_155999869_A_G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>rs3091261</td>\n",
       "      <td>[1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128182</th>\n",
       "      <td>chrX_155999957_G_C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>rs3093472</td>\n",
       "      <td>[1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128183</th>\n",
       "      <td>chrX_156025297_G_A</td>\n",
       "      <td>downstream</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>rs185932868</td>\n",
       "      <td>[0/1, 0/0, 0/1, 0/0, 0/1, 0/0, 0/0, 0/1, 0/1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128187</th>\n",
       "      <td>chrY_3120449_C_G</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>LINC00278;TGIF2LY</td>\n",
       "      <td>rs778218266</td>\n",
       "      <td>[0/1, 1/1, 0/1, 1/1, 0/0, 1/1, 1/1, 0/0, 0/1, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127155 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variants      function              genes        avsnp  \\\n",
       "0           chr1_14574_A_G  ncRNA_exonic             WASH7P   rs28503599   \n",
       "1           chr1_14590_G_A  ncRNA_exonic             WASH7P     rs707679   \n",
       "2           chr1_14599_T_A  ncRNA_exonic             WASH7P     rs707680   \n",
       "3           chr1_14604_A_G  ncRNA_exonic             WASH7P  rs541940975   \n",
       "4           chr1_14610_T_C  ncRNA_exonic             WASH7P  rs878986575   \n",
       "...                    ...           ...                ...          ...   \n",
       "128180  chrX_155999818_C_T      intronic               IL9R    rs3093470   \n",
       "128181  chrX_155999869_A_G      intronic               IL9R    rs3091261   \n",
       "128182  chrX_155999957_G_C      intronic               IL9R    rs3093472   \n",
       "128183  chrX_156025297_G_A    downstream           DDX11L16  rs185932868   \n",
       "128187    chrY_3120449_C_G    intergenic  LINC00278;TGIF2LY  rs778218266   \n",
       "\n",
       "                                                       GT  \n",
       "0       [0/0, 0/1, 0/1, 0/0, 0/0, 0/0, 0/1, 0/0, 0/1, ...  \n",
       "1       [0/0, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/0, 0/0, ...  \n",
       "2       [0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...  \n",
       "3       [0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...  \n",
       "4       [0/1, 0/1, 0/1, 0/0, 0/0, 0/1, 0/1, 0/1, 0/1, ...  \n",
       "...                                                   ...  \n",
       "128180  [1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...  \n",
       "128181  [1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, 1/1, ...  \n",
       "128182  [1/1, 0/0, 1/1, 1/1, 0/1, 0/1, 0/1, 1/1, 1/1, ...  \n",
       "128183  [0/1, 0/0, 0/1, 0/0, 0/1, 0/0, 0/0, 0/1, 0/1, ...  \n",
       "128187  [0/1, 1/1, 0/1, 1/1, 0/0, 1/1, 1/1, 0/0, 0/1, ...  \n",
       "\n",
       "[127155 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_unique=np.unique(df_d[\"genes\"].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_unique=np.unique(df_d[\"genes\"].tolist())\n",
    "a=np.array(df_b[\"AD\"])\n",
    "d=np.array(df_b[\"DP\"])\n",
    "a=np.array(list(map(float,a)))\n",
    "d=np.array(list(map(float,d)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "model =vireoSNP.VireoBulk(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.fit(a,d,m_GT,learn_theta=True )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.01164797, 0.44852633, 0.99267153])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.theta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['18-G-017_R1',\n",
       " '18-G-018_R1',\n",
       " '18-G-019_R1',\n",
       " '18-G-020_R1',\n",
       " '18-G-021_R1',\n",
       " '18-G-022_R1',\n",
       " '18-G-023_R1',\n",
       " '18-G-024_R1',\n",
       " '18-G-025_R1',\n",
       " '18-G-026_R1']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "donor_vcf['samples']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.06418119, 0.04299516, 0.05283177, 0.08967812, 0.14387843,\n",
       "       0.06906367, 0.04872266, 0.11876265, 0.29342823, 0.07645813])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.psi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "premodel=model.psi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "###gene model\n",
    "\n",
    "bulk_demuti_result= prediction_bulk(df_b,df_d,m_GT,donor_vcf,premodel,gene_unique)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>18-G-017_R1</th>\n",
       "      <th>18-G-018_R1</th>\n",
       "      <th>18-G-019_R1</th>\n",
       "      <th>18-G-020_R1</th>\n",
       "      <th>18-G-021_R1</th>\n",
       "      <th>18-G-022_R1</th>\n",
       "      <th>18-G-023_R1</th>\n",
       "      <th>18-G-024_R1</th>\n",
       "      <th>18-G-025_R1</th>\n",
       "      <th>18-G-026_R1</th>\n",
       "      <th>chi</th>\n",
       "      <th>p</th>\n",
       "      <th>numSNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.240487</td>\n",
       "      <td>0.031524</td>\n",
       "      <td>0.246912</td>\n",
       "      <td>0.0424</td>\n",
       "      <td>0.212023</td>\n",
       "      <td>0.01513</td>\n",
       "      <td>0.073156</td>\n",
       "      <td>0.081881</td>\n",
       "      <td>0.022049</td>\n",
       "      <td>0.034438</td>\n",
       "      <td>3.290594</td>\n",
       "      <td>0.951652</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.322538</td>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.043626</td>\n",
       "      <td>0.000298</td>\n",
       "      <td>0.024924</td>\n",
       "      <td>0.110349</td>\n",
       "      <td>0.003099</td>\n",
       "      <td>0.307019</td>\n",
       "      <td>0.003918</td>\n",
       "      <td>0.183691</td>\n",
       "      <td>2.721691</td>\n",
       "      <td>0.974315</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.111379</td>\n",
       "      <td>0.186228</td>\n",
       "      <td>0.015524</td>\n",
       "      <td>0.070369</td>\n",
       "      <td>0.002524</td>\n",
       "      <td>0.00208</td>\n",
       "      <td>0.113103</td>\n",
       "      <td>0.210372</td>\n",
       "      <td>0.260654</td>\n",
       "      <td>0.027765</td>\n",
       "      <td>6.853791</td>\n",
       "      <td>0.652339</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M;PZP</th>\n",
       "      <td>0.297699</td>\n",
       "      <td>0.057649</td>\n",
       "      <td>0.159085</td>\n",
       "      <td>0.049461</td>\n",
       "      <td>0.025333</td>\n",
       "      <td>0.013277</td>\n",
       "      <td>0.182469</td>\n",
       "      <td>0.030398</td>\n",
       "      <td>0.16116</td>\n",
       "      <td>0.023469</td>\n",
       "      <td>5.948257</td>\n",
       "      <td>0.745086</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2MP1</th>\n",
       "      <td>0.054614</td>\n",
       "      <td>0.433896</td>\n",
       "      <td>0.016346</td>\n",
       "      <td>0.109016</td>\n",
       "      <td>0.105064</td>\n",
       "      <td>0.050424</td>\n",
       "      <td>0.014692</td>\n",
       "      <td>0.170016</td>\n",
       "      <td>0.045928</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>22.90447</td>\n",
       "      <td>0.006414</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>0.183335</td>\n",
       "      <td>0.069798</td>\n",
       "      <td>0.073204</td>\n",
       "      <td>0.156388</td>\n",
       "      <td>0.011695</td>\n",
       "      <td>0.013849</td>\n",
       "      <td>0.084582</td>\n",
       "      <td>0.1278</td>\n",
       "      <td>0.164845</td>\n",
       "      <td>0.114505</td>\n",
       "      <td>2.391586</td>\n",
       "      <td>0.983658</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.084576</td>\n",
       "      <td>0.162609</td>\n",
       "      <td>0.022348</td>\n",
       "      <td>0.183719</td>\n",
       "      <td>0.0511</td>\n",
       "      <td>0.027481</td>\n",
       "      <td>0.088382</td>\n",
       "      <td>0.223326</td>\n",
       "      <td>0.086492</td>\n",
       "      <td>0.069967</td>\n",
       "      <td>10.147248</td>\n",
       "      <td>0.338697</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0.026771</td>\n",
       "      <td>0.061399</td>\n",
       "      <td>0.018724</td>\n",
       "      <td>0.008064</td>\n",
       "      <td>0.084577</td>\n",
       "      <td>0.122853</td>\n",
       "      <td>0.104039</td>\n",
       "      <td>0.288391</td>\n",
       "      <td>0.283356</td>\n",
       "      <td>0.001826</td>\n",
       "      <td>7.684969</td>\n",
       "      <td>0.566179</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.063709</td>\n",
       "      <td>0.064758</td>\n",
       "      <td>0.008241</td>\n",
       "      <td>0.103454</td>\n",
       "      <td>0.050555</td>\n",
       "      <td>0.040779</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.09356</td>\n",
       "      <td>0.474027</td>\n",
       "      <td>0.100855</td>\n",
       "      <td>1.673724</td>\n",
       "      <td>0.995641</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.036555</td>\n",
       "      <td>0.076785</td>\n",
       "      <td>0.262047</td>\n",
       "      <td>0.230531</td>\n",
       "      <td>0.154715</td>\n",
       "      <td>0.06554</td>\n",
       "      <td>0.002896</td>\n",
       "      <td>0.050482</td>\n",
       "      <td>0.036244</td>\n",
       "      <td>0.084205</td>\n",
       "      <td>4.397809</td>\n",
       "      <td>0.883336</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13350 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         18-G-017_R1 18-G-018_R1 18-G-019_R1 18-G-020_R1 18-G-021_R1  \\\n",
       "A1BG        0.240487    0.031524    0.246912      0.0424    0.212023   \n",
       "A1BG-AS1    0.322538    0.000538    0.043626    0.000298    0.024924   \n",
       "A2M         0.111379    0.186228    0.015524    0.070369    0.002524   \n",
       "A2M;PZP     0.297699    0.057649    0.159085    0.049461    0.025333   \n",
       "A2MP1       0.054614    0.433896    0.016346    0.109016    0.105064   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZXDC        0.183335    0.069798    0.073204    0.156388    0.011695   \n",
       "ZYG11B      0.084576    0.162609    0.022348    0.183719      0.0511   \n",
       "ZYX         0.026771    0.061399    0.018724    0.008064    0.084577   \n",
       "ZZEF1       0.063709    0.064758    0.008241    0.103454    0.050555   \n",
       "ZZZ3        0.036555    0.076785    0.262047    0.230531    0.154715   \n",
       "\n",
       "         18-G-022_R1 18-G-023_R1 18-G-024_R1 18-G-025_R1 18-G-026_R1  \\\n",
       "A1BG         0.01513    0.073156    0.081881    0.022049    0.034438   \n",
       "A1BG-AS1    0.110349    0.003099    0.307019    0.003918    0.183691   \n",
       "A2M          0.00208    0.113103    0.210372    0.260654    0.027765   \n",
       "A2M;PZP     0.013277    0.182469    0.030398     0.16116    0.023469   \n",
       "A2MP1       0.050424    0.014692    0.170016    0.045928    0.000005   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZXDC        0.013849    0.084582      0.1278    0.164845    0.114505   \n",
       "ZYG11B      0.027481    0.088382    0.223326    0.086492    0.069967   \n",
       "ZYX         0.122853    0.104039    0.288391    0.283356    0.001826   \n",
       "ZZEF1       0.040779    0.000061     0.09356    0.474027    0.100855   \n",
       "ZZZ3         0.06554    0.002896    0.050482    0.036244    0.084205   \n",
       "\n",
       "                chi         p numSNP  \n",
       "A1BG       3.290594  0.951652      3  \n",
       "A1BG-AS1   2.721691  0.974315      2  \n",
       "A2M        6.853791  0.652339      2  \n",
       "A2M;PZP    5.948257  0.745086      1  \n",
       "A2MP1      22.90447  0.006414      3  \n",
       "...             ...       ...    ...  \n",
       "ZXDC       2.391586  0.983658     13  \n",
       "ZYG11B    10.147248  0.338697      9  \n",
       "ZYX        7.684969  0.566179      2  \n",
       "ZZEF1      1.673724  0.995641     29  \n",
       "ZZZ3       4.397809  0.883336     25  \n",
       "\n",
       "[13350 rows x 13 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###the 10 donor gene level demutiplexing may not as accurate as it in two donors  this result is just for demonstration\n",
    "bulk_demuti_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
